torch/csrc/jit/mobile/backport_manager.cpp

```diff
@@ -211,6 +212,152 @@ std::stringstream backport_v5_to_v4(std::stringstream& input_model_stream) {
   return ouput_model_stream;
 }
 
+void writeArchiveV5(
+    PyTorchStreamWriter& writer,
+    const IValue& value,
+    const std::string& archive_name,
+    const std::string& archive_dir,
+    const std::string& tensor_dir,
+    bool tensor_cdata_naming_scheme) {
+  std::vector<char> data;
+  // Vector to capture the run-time class types during pickling the IValues
+  std::vector<c10::ClassTypePtr> memoizedClassTypes;
+  std::vector<std::string> tensor_names;
+  Pickler data_pickle(
+      [&](const char* buf, size_t size) {
+        data.insert(data.end(), buf, buf + size);
+      },
+      nullptr,
+      nullptr,
+      &memoizedClassTypes,
+      [&](const at::Tensor& tensor) {
+        // returns a string to use in picker.cpp as storage obj key
+        if (tensor_cdata_naming_scheme) {
+          tensor_names.push_back(
+              std::to_string(reinterpret_cast<std::intptr_t>(
+                  tensor.storage().unsafeGetStorageImpl())) +
+              ".storage");
+        } else {
+          tensor_names.push_back(std::to_string(tensor_names.size()));
+        }
+        return tensor_names.back();
+      });
+  data_pickle.protocol();
+  data_pickle.pushIValue(value);
+  data_pickle.stop();
+  // write out tensor data
+  size_t i = 0;
+  std::string prefix = archive_name + "/";
+
+  TORCH_INTERNAL_ASSERT(tensor_names.size() == data_pickle.tensorData().size());
+  const std::vector<std::string>& pre_serialized_files =
+      writer.getAllWrittenRecords();
+
+  for (const auto& td : data_pickle.tensorData()) {
+    WriteableTensorData writable_td = getWriteableTensorData(td);
+    std::string fname = tensor_dir + tensor_names[i++];
+    if (tensor_cdata_naming_scheme &&
+        std::find(
+            pre_serialized_files.begin(), pre_serialized_files.end(), fname) !=
+            pre_serialized_files.end()) {
+      // storage has been serialzed already, skip
+      continue;
+    }
+    writer.writeRecord(fname, writable_td.data(), writable_td.sizeInBytes());
+  }
+
+  std::string fname = archive_dir + archive_name + ".pkl";
+  writer.writeRecord(fname, data.data(), data.size());
+
+  // serialize all the captured run-time class types
+  //  for (const c10::ClassTypePtr& wroteType : memoizedClassTypes) {
+  //    convertNamedType(wroteType);
+  //  }
+}
+
+std::stringstream backport_v6_to_v5(std::stringstream& input_model_stream) {
```

Comment:
Implementation looks good, however I have a high level question: why do we backport byte code and constant table when we can re-generate an entirely new archive that is v5 compatible?

Reply:
Sorry I missed this question. The purpose for backport is for forward compatibility: to deploy a new model (generated by new code) to old runtime. The reason we use backport function, is to keep one unified serialization version for current code to avoid confusion. If we re-generate an entirely new archive for v5, that means that we maintain serialization flows to both v6 and v5 in the main code. Another reason is to export to even older versions, that the backport functions can be called in a while loop. In general the backport function should be just format change. 

It's special case for this v6 update, where the bytecode must be remitted to backport, that changing format of bytecode is not possible.

Reply:
>  If we re-generate an entirely new archive for v5, that means that we maintain serialization flows to both v6 and v5 in the main code. 

By maintaining backport code, it is pretty much the same as maintaining both v5 and v6 export in main code.

> Another reason is to export to even older versions, that the backport functions can be called in a while loop. In general the backport function should be just format change.

Do we need to backport to older than v5 (like v4)?

Reply:
To be clear, I think current approach works too, just curious why it is done this way.